azkaban-common/src/main/java/azkaban/executor/OnContainerizedExecutionEventListener.java

```diff
@@ -70,22 +70,44 @@ private void restartExecutableFlow(final ExecutableFlow exFlow) {
     // Update the flow options so that the flow will be not retried again by Azkaban
 
     // inherent the retry time counters
-    executableFlow.setUserDefinedRetryCount(exFlow.getUserDefinedRetryCount());
-    executableFlow.setSystemDefinedRetryCount(exFlow.getSystemDefinedRetryCount());
+    retryExFlow.setUserDefinedRetryCount(originalExFlow.getUserDefinedRetryCount());
+    retryExFlow.setSystemDefinedRetryCount(originalExFlow.getSystemDefinedRetryCount());
+
+    if (originalExFlow.getFlowRetryRootExecutionID() > 0) {
+      retryExFlow.setFlowRetryRootExecutionID(originalExFlow.getFlowRetryRootExecutionID());
+    } else {
+      retryExFlow.setFlowRetryRootExecutionID(originalExFlow.getExecutionId());
+    }
+    retryExFlow.setFlowRetryParentExecutionID(originalExFlow.getExecutionId());
 
     // If a retried flow A gets retried again with a new execution id flow B, the original flow
     // execution id of flow B should be the same as flow A's original flow execution id.
     if (options.getOriginalFlowExecutionIdBeforeRetry() == null) {
-      options.setOriginalFlowExecutionIdBeforeRetry(exFlow.getExecutionId());
+      options.setOriginalFlowExecutionIdBeforeRetry(originalExFlow.getExecutionId());
     }
-    executableFlow.setExecutionOptions(options);
+    retryExFlow.setExecutionOptions(options);
     // Submit new flow for execution
     try {
-      logger.info("Restarting flow " + project.getName() + "." + executableFlow.getFlowName());
-      this.executorManagerAdapter.submitExecutableFlow(executableFlow,
-          executableFlow.getSubmitUser());
+      logger.info("Restarting flow " + project.getName() + "." + retryExFlow.getFlowName());
```

Comment:
log with number of retries, e.g. restarting flow *** with {[0~9]}th retry, the maxinum of auto retry is {}.

Reply:
good point! let me add